[Tan Chee Heng] iP

cheeheng opened the pull request:
## Jude

> Hey Jude, don't make it bad.
> Take a sad song and make it better. - lyrics of pop song hit [_Hey Jude_](https://www.youtube.com/watch?v=A_MjCqQoLLA), The Beatles (1968)

Fast forward to 2022, and Jude the chatbot is born. Jude is a
- Text-based command-line interface
- Task tracker which helps track events and deadlines

To use Jude the chatbot, do the following:
1. Download the file Jude-0.1.jar [here](https://github.com/cheeheng/ip/releases)
1. Type the command `java -jar Jude-0.1.jar` on the ~different~ **same** directory as Jude-0.1.jar
1. Add tasks
1. Jude the chatbot will help you manage the tasks 😊

Here are some features of Jude the chatbot:
- [x] Add and delete tasks
   - [x] All tasks have a description
   - [x] Todo tasks
   - [x] Deadline tasks have a deadline which needs to be a proper date
   - [x] Event tasks have a start and end time which is currently condensed under 1 parameter /at 
- [ ] Sort deadline tasks by deadline (not yet implemented)
- [ ] Event tasks have a separate start and end time (not yet implemented)

This is the ```main()``` method which was used in an earlier version of the chatbot:
```java
/**
 * Runs the task tracker chatbot.
 *
 * @param args Not used for now.
 */
public static void main(String[] args) throws IOException {
    new Jude("data/tasks.txt").run();
}
```

Commit:
Add Gradle support

cheeheng committed:
Level 1: Greet, Echo, Exit - renamed chatbot to Jude, turned chatbot into something like command line interface.

cheeheng committed:
The command to exit should be bye instead of exit.

cheeheng committed:
Level 2: Add list functionality.

cheeheng committed:
Add new Task.java file

cheeheng committed:
Level 3: Support marking task as done.

cheeheng committed:
Add class documentation to Jude and Task

cheeheng committed:
Create Todo class.

cheeheng committed:
Add Deadline and Event classes as descendants of Task.

cheeheng committed:
Add getter method for deadline object in Deadline.

cheeheng committed:
Level 4: Add support for multiple types of tasks

cheeheng committed:
A-TextUiTesting: Add test

cheeheng committed:
Level 5: Handle errors due to invalid input

cheeheng committed:
Level 6: Allow deletion of tasks.

cheeheng committed:
Add load and save functionality

Right now, the save functionality is not present, which means that the
data in the task tracker is lost once the program stops running.

Let's add the save functionality to preserve the state of the task
tracker across different runs.

cheeheng committed:
Add tests to check save functionality

cheeheng committed:
Add tests to ensure completion marker saved correctly

cheeheng committed:
Add proper date checking to Jude

cheeheng committed:
Merge branch 'master' into branch-Level-8

cheeheng committed:
Revert "Add proper date checking to Jude"

This reverts commit c200492c86744eaf06e03aa13b9b9ae5bb140061.

cheeheng committed:
Merge branch 'branch-Level-7'

cheeheng committed:
Merge branch 'branch-Level-8'

cheeheng committed:
Revert "Revert "Add proper date checking to Jude""

This reverts commit 7df675c2875dfe77165f028c0e319744f4a9fd32.

cheeheng committed:
Revert "Revert "Add proper date checking to Jude""

This reverts commit 7df675c2875dfe77165f028c0e319744f4a9fd32.

(cherry picked from commit 1bfafdddeb330f8023b7fdd438e2791fbe0a886a)

cheeheng committed:
Fix EXPECTED.TXT

cheeheng committed:
Revert "Revert "Revert "Add proper date checking to Jude"""

This reverts commit 1bfafdddeb330f8023b7fdd438e2791fbe0a886a.

cheeheng committed:
Merge branch 'branch-Level-8'

cheeheng committed:
Revert "Revert "Revert "Revert "Add proper date checking to Jude""""

This reverts commit 8cf47bbc2c3fc668763533eb017ad930c0d04295.

cheeheng committed:
Fix EXPECTED.TXT

cheeheng committed:
Abstract file storage and task list functionality

cheeheng committed:
Refactor code to OOP style

cheeheng committed:
Remove unnecessary imports

cheeheng committed:
Organize code into packages

cheeheng committed:
Merge branch 'add-gradle-support'

Merge branch 'add-gradle-support' of
https://github.com/nus-cs2103-AY2223S1/ip

cheeheng committed:
Add JUnit tests

cheeheng committed:
Configure build.gradle for v0.1 release

cheeheng committed:
Refine JavaDoc documentation

(cherry picked from commit 89c24dd278f9106414a5cae49ec2e2505daf9190)

cheeheng committed:
Refine JavaDoc documentation

cheeheng committed:
Fix code lint issues excluding JavaDoc

cheeheng committed:
Add search functionality to Jude

cheeheng committed:
Merge branch 'branch-Level-9'

cheeheng committed:
Merge branch 'branch-A-JavaDoc'

cheeheng committed:
Merge branch 'branch-A-CodingStandard'

cheeheng committed:
Add Checkstyle to project and fix Checkstyle errors

cheeheng committed:
Add JavaFX GUI to Jude the chatbot

cheeheng committed:
Add JavaFX GUI to Jude the chatbot

(cherry picked from commit eeffb6a066bedc53e4718f07dd15257b65361599)

cheeheng committed:
Fix bugs after moving code to JavaFX

cheeheng committed:
Fix bugs after moving code to JavaFX

(cherry picked from commit 27025071bc11cda8e69c10819eca0b716dd56258)

cheeheng committed:
Fix empty line JavaDoc coding standard violation

cheeheng committed:
Merge branch 'branch-Level-10'

# Conflicts:
#	src/main/java/jude/DialogBox.java
#	src/main/java/jude/Jude.java
#	src/main/java/jude/Launcher.java
#	src/main/java/jude/Main.java
#	src/main/java/jude/MainWindow.java
#	src/main/java/jude/Parser.java

cheeheng committed:
Fix bugs in chatbot GUI

Josephaedan commented on src/main/java/jude/Main.java:
```diff
@@ -0,0 +1,48 @@
+ * A GUI for Jude the chatbot using FXML.
+ */
+public class Main extends Application {
+    private final Jude jude = new Jude("data/tasks.txt");
```
For easier modification, perhaps you could consider refactoring 'data/tasks.txt' into a constant TASKS_LIST or something outside the code just for easier readability/editing in the future!

Josephaedan commented on src/main/java/jude/Parser.java:
```diff
@@ -0,0 +1,281 @@
+        command = command.trim(); // to remove unintended trailing spaces
+        String[] tokens = command.split(" ", 2);
+        String response = "";
+        try {
```
You wrote this function to parse and execute the command provided. However, as a result, it is really long and difficult to read the different cases. Perhaps you could consider refactoring out the different execution codes to separate functions so that it is much cleaner and readable.

cheeheng replied:
I agree with this. In addition, line 117 contains a redundant comment which I should remove.

Josephaedan commented on src/main/java/jude/task/Task.java:
```diff
@@ -0,0 +1,110 @@
+    /**
+     * Sets the status of the {@code Task} to 'done'.
+     */
+    public void markAsDone() {
```
It seems like markAsDone(), markAsUndone() and setIsDone(isDone) all serve the purpose of setting this.isDone property. I'm not really sure about the value of setIsDone() being executed in other parts of your codebase as this is bad practice; Ideally it should only be called within the class itself. Perhaps you could consider just removing setIsDone() and using markAsDone() and markAsUndone() for the rest of the codebase. That way you can do proper error handling within the functions themselves too.

Josephaedan reviewed (commented):
Overall, I think your code looks good! My only concern is for people reviewing your code in the future, understanding the parsing and executing of the command may be a little too tedious and difficult so you could consider abstracting them into separate functions.

cheeheng commented:
Thank you for your input. Yes, I agree that the parsing code is a bit too long and I would probably have to refactor this part into several functions.

cheeheng committed:
Update build.gradle to comply with reuse policy

cheeheng committed:
Restore regression testing for Gradle project

The old regression test had been removed because the program has been
converted from a command-line interface to a graphical user interface.

A lack of regression testing may introduce bugs as new features are
being added.

Let's modify the runtest.bat script and restore the methods intended for
printing to console to support console-based regression testing.

The regression testing is restored because it will be useful for testing
while adding new features to the graphical user interface. This helps to
verify that no functionality has been accidentally changed as a result
of adding in new features.

cheeheng committed:
Add assertions to code

There are no assertions in the code.

When there is a NullPointerException, the code could be hard to debug
because it can happen almost anywhere.

Let's add assertions to the code to aid debugging when variables
stored become null.

When there is a JavaFX invocation exception or similar, these can be
hard to debug. Occasionally, that could be caused by null values. By
catching this earlier, we would know which part of the code was not
working correctly by throwing up null values.

cheeheng committed:
Refactor parser method

The parser code is cluttered into one method, with the parser method
parsing the commands and also executing them, hence making the code hard
to read.

Let's refactor the parser code to better adhere to the single level of
abstraction principle.

This will make the code easier to read, as suggested by a reviewer of my
code (Josephaedan).

This commit also includes other minor code changes.

cheeheng committed:
Improve code quality with minor corrections

cheeheng committed:
Merge pull request #1 from cheeheng/branch-A-CodeQuality

Code Quality Improvements

cheeheng committed:
Merge branch 'master' into branch-A-Assertions

# Conflicts:
#	src/main/java/jude/DialogBox.java
#	src/main/java/jude/Launcher.java
#	src/main/java/jude/Parser.java

cheeheng committed:
Merge pull request #2 from cheeheng/branch-A-Assertions

Add assertions to code

cheeheng committed:
Setup Continuous Integration

cheeheng committed:
Fix broken CI

cheeheng committed:
Fix broken CI: Attempt 2

cheeheng committed:
Fix broken CI: Attempt 3

cheeheng committed:
Fix broken CI: Attempt 4

cheeheng committed:
Fix gradlew directory issue in runtest.sh

cheeheng committed:
Add reminder feature to Jude the chatbot

yuxuanleong commented on src/main/java/jude/Storage.java:
```diff
@@ -0,0 +1,113 @@
+            boolean isDone = Integer.parseInt(done) == 1;
+
+            Task task;
+            if (taskType.equals("T")) {
+                task = new Todo(taskName, isDone);
+            } else if (taskType.equals("D")) {
+                String deadline = sc.nextLine();
+                task = new Deadline(taskName, isDone, deadline);
+            } else if (taskType.equals("E")) {
+                String eventTime = sc.nextLine();
+                task = new Event(taskName, isDone, eventTime);
```
This can probably be made into a method on itself and the code will look cleaner 😄

yuxuanleong reviewed (commented):
You are generally doing a really good job, just some tidying up jobs for longer methods!

cheeheng committed:
Fix code quality issues detected by soc-se-bot

cheeheng committed:
Increase margin between text and profile image

cheeheng committed:
Remove credits message because there is separate file for credits

cheeheng committed:
Merge pull request #4 from cheeheng/branch-B-Reminders

Merge Reminders branch into master

cheeheng committed:
Add support for resizing GUI

cheeheng committed:
Edit comments to better comply with reuse policy

cheeheng committed:
Add support for event dates in Jude the chatbot

cheeheng committed:
Fix checkstyle error

cheeheng committed:
Move credits to root directory to make it more obvious

cheeheng committed:
Set message box to adjust the size according to window size

cheeheng committed:
Fix bug where window resize causes text to not render correctly

cheeheng committed:
Include credit for idea on adding start and end time for event

cheeheng committed:
Add Ui.png for iP showcase

cheeheng committed:
Update README.md in root directory

cheeheng committed:
Fix regression where reminder text is missing in GUI version

cheeheng committed:
Modify Ui.png to reflect GUI change

cheeheng committed:
Update documentation except for usage section

cheeheng committed:
Update user guide to first complete draft

cheeheng committed:
Update event to disallow start time after end time

cheeheng committed:
Release v0.3

cheeheng committed:
Refactor error messages to constants

cheeheng committed:
Update user documentation to include credits and installation steps

cheeheng committed:
Change fxml files to version 11 to resolve warning

cheeheng committed:
Add link to user guide on README.md (project root directory)

cheeheng committed:
Extract magic constant from invalid index error message

cheeheng committed:
Standardize @@author reuse tags

cheeheng committed:
Fix QuantumRenderer error

cheeheng committed:
Fix bug where delete given without arguments

cheeheng committed:
Fix picture does not show bug in Linux

cheeheng committed:
Fix picture not showing (Attempt 2)

cheeheng committed:
Fix Wayland incompatibility

cheeheng committed:
Fix delete non-numeric bug

cheeheng committed:
Fix minor bugs in user guide
